### PR TITLE
[SPARK-25765][ML] Add training cost to BisectingKMeans summary

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The PR adds the `trainingCost` value to the `BisectingKMeansSummary`, in order to expose the information retrievable by running `computeCost` on the training dataset. This fills the gap with `KMeans` implementation.

## How was this patch tested?

improved UTs

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @cloud-fan @dongjoon-hyun @holdenk @srowen

**Comment by cloud-fan** [comment]
does the example need to be updated with this new API?

**Comment by mgaido91** [comment]
For KMeans we used the ClusteringEvaluator in the examples. Actually, the training cost is not a good way to evaluate a dataset (the evaluation should be done on a dataset different from the training one).

Maybe we can also add this API in the example to show that it exists, but I have seen no example with a summary shown so...

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala [hunk 1]
```diff
@@ -310,4 +317,6 @@ class BisectingKMeansSummary private[clustering] (
     predictionCol: String,
     featuresCol: String,
     k: Int,
-    numIter: Int) extends ClusteringSummary(predictions, predictionCol, featuresCol, k, numIter)
+    numIter: Int,
+    @Since("2.4.0") val trainingCost: Double)
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: 2.4.0? or 3.0.0?

Comment (cloud-fan) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: this PR targets to 2.4, see more context at https://github.com/apache/spark/pull/22756

Comment (cloud-fan) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: oh wait. If the final goal is to have a consistent ML API in 3.0, do we have to put this new API in 2.4?

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: I wouldn't consider this as mandatory. I think what is mandatory to target for 2.4 is to deprecate the `computeCost` method. But I think it is a nice to have, since it is a non-deprecated way users have to access this information. In the PR related to `KMeans`, there was quite a discussion about it and it was considered to be part of the deprecation change.

Comment (cloud-fan) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: SGTM

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: Ok. I see. Then looks like it is nice to have this in 2.4.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala [hunk 2]
```diff
@@ -225,13 +227,14 @@ object BisectingKMeansModel extends Loader[BisectingKMeansModel] {
       assert(formatVersion == thisFormatVersion)
       val rootId = (metadata \ "rootId").extract[Int]
       val distanceMeasure = (metadata \ "distanceMeasure").extract[String]
+      val trainingCost = (metadata \ "trainingCost").extract[Double]
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: hmm, can this read old model from previous version?

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: - Could you avoid modifying loading model code in "mllib" package, but modifying code in "ml" package, i.e., the class `ml.clustering.BisectingKMeansModel.BisectingKMeansModelReader`, you can reference the `KMeans` code: `ml.clustering.KMeansModel.KMeansModelReader`.

- And, +1 with @viirya mentioned, we should keep model loading compatibility, add a version check (when >= 2.4) then we load "training cost" . Note that add these in `ml.clustering.BisectingKMeansModel.BisectingKMeansModelReader`.

- And, could you also add version check (when >= 2.4) then we load "training cost" into `ml.clustering.KMeansModel.KMeansModelReader` ?

**Review by WeichenXu123 — CHANGES_REQUESTED** [review]
Several suggestion. Thanks!

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 2 above)
Comment (cloud-fan) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Do other models have this problem? I was told that this change just follows what we did for other models before.

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Thank you all for the comments and sorry for the late answer. Just a couple of notes on your comments @WeichenXu123  (I may be missing something, so please correct me if I am wrong):

 - I checked the `ml.clustering.KMeansModel.KMeansModelReader` and it doesn't store anything related to the summary. Summary is not recovered after load of the model, so I don't see any reason why we should modify the read/load of `ml.clustering.BisectingKMeansModel.BisectingKMeansModelReader`;
 - this model can read from previous versions, since this is version "2.0", which was introduced for Spark 2.4; for previous versions, we read/write version "1.0"; the version check method for versioning is used only for the `ml` package, not in `mllib` where we have this versioning approach;

> I was told that this change just follows what we did for other models before.

@cloud-fan Yes, let me link the PR for KMeans doing the same, which is: https://github.com/apache/spark/pull/20629.

Just a final comment which I hope clarifies which is the source of the confusion here and the reason of the above comments by @viirya and @WeichenXu123: `trainingCost` here is a member of the `summary`, not a parameter of the model for the `ml.clustering.BisectingKMeansModel`. Instead, it is a member of the model for `mllib.clustering.BisectingKMeansModel` (we have no summary notion there...). So storing it for `mllib` is needed in order for the model read after persisting it being the same of the original one (I think it doesn't pass UTs otherwise). Storing it for the `ml`, instead, it is not needed, because the summary is not persisted. If we want to persist also the summary for `ml` package I think we should best create a separate JIRA and PR for it.

Hope this clarifies (sorry for being so verbose).

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: > this model can read from previous versions, since this is version "2.0", which was introduced for Spark 2.4; for previous versions, we read/write version "1.0"; the version check method for versioning is used only for the ml package, not in mllib where we have this versioning approach;

I meant that can it read old model from previous versions, not that this model can read from previous versions.

In other words, when reading a previous model without "trainingCost" in metadata, can this line work well?

```scala
val trainingCost = (metadata \ "trainingCost").extract[Double]
```

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: The answer is yes, because of versioning:
 - all the versions of Spark released so far, use the versions `"1.0"`, which is loaded/written with the `SaveLoadV1_0` class (see https://github.com/apache/spark/pull/22764/files#diff-458b0e49bfc4a4dd993ea6cb66af4b4dR125);
 - there can't be a model with version `"2.0"` without `trainingCost`, because `SaveLoadV2_0` was introduced in a previous commit which lands only on 2.4 and has not yet been released.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: Since #22756 is reverted, are we going to change the `Since` version for this, too?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 2 above)
Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Ok. I see. `SaveLoadV2_0` is confused term as I think it is for reading models since Spark 2.0.

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: @mgaido91 
What about this way ?
In `ml` reader, we can load `trainingCost` and then construct `mllib.clustering.BisectingKMeansModel` and pass the `trainingCost` argument and then construct `ml.clustering.BisectingKMeansModel` to wrap that `mllib` model.
** Then in `ml` reader we can check the spark major/minor version to keep backwards compatibility, this is a more important thing. **
** In `mllib` loader we can ignore loading `trainingCost`, because `mllib` is deprecate, we can ignore adding new features, but we cannot breaking backwards compatibility (your change breaking backwards compatibility). **

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: @WeichenXu123 I don't see (and think) this change breaks backwards compatibility for `mllib`. You can still load a model stored by previous versions, the only difference is that it won't have a valid `trainingCost` (but this just means we lack a new feature, all operations/data available before is still there, so backward compatibility is ensured). The same approach was used also for `KMeans`, so if we want to change approach here, we might have to do it also for `KMeans`.

Anyway, regarding your suggestion, since the summary is not recomputed, there is no need at all to store this for `ml`: a loaded model doesn't contain the summary, so doesn't contain the `trainingCost` (same for `KMeans`). We could/should rather ignore at all storing the value, it changes nothing on `ml` side.

On the other side, if we want it to be stored and re-loaded on `ml` side, we should start persisting the summary too there, but as I mentioned before, I think this is a different and bigger change, which should involve all the summary/models and not only `BisectingKMeans`.

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: @mgaido91 
(I haven't test this, so correct me if I am wrong).
> I don't see (and think) this change breaks backwards compatibility for mllib.

I am suspicious of this line:
```
val trainingCost = (metadata \ "trainingCost").extract[Double]
```
When loading an old version spark saved `BisectingKMeansModel`, because it do not contain "trainingCost" info, I guess this line will throw error. (Otherwise what will it return ?)

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: yes @WeichenXu123 , you're right, that line is a bug. Thanks for noticing it. Anyway, that is going to be addressed in another PR and it is not (strictly) related to this one. The other option, as I mentioned, is that if we agree that this doesn't need to be restored after model persistence, we can just ignore it in save/load.

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: OK. After https://github.com/apache/spark/pull/22790 merged, I think this PR can work.

**Comment by mgaido91** [comment]
as @WeichenXu123 mentioned in https://github.com/apache/spark/pull/22764#discussion_r228527916, I don't see other problems with the current PR. The only thing is: do we want to target it for 2.4 or for 3.0? If the latter, I'll update the PR with the proper deprecation messages. Thanks.

**Comment by WeichenXu123** [comment]
I think it can target for 3.0. since 2.4 will be released soon and this PR looks a little complex and need take  some time to check carefully.

**Comment by mgaido91** [comment]
thanks @WeichenXu123 , I updated this PR in order to target 3.0.

**Comment by mgaido91** [comment]
any more comments on this? Thanks

**Comment by mgaido91** [comment]
@dbtsai any comments on this? thanks.

**Comment by dbtsai** [comment]
@mgaido91 I'm on thanksgiving vacation, will be back to community to help code review on Nov 21st. Sorry for the delay.

**Comment by mgaido91** [comment]
@dbtsai sure, thanks. Sorry for bothering you. Have a nice vacation!

**Comment by mgaido91** [comment]
@dbtsai any luck with this? Thanks.

**Comment by mgaido91** [comment]
kindly ping @dbtsai

**Comment by mgaido91** [comment]
since this has been stuck for a while, maybe @holdenk @srowen  can you help reviewing this? thanks.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala [hunk 3]
```diff
@@ -302,6 +307,8 @@ object BisectingKMeans extends DefaultParamsReadable[BisectingKMeans] {
  * @param featuresCol  Name for column of features in `predictions`.
  * @param k  Number of clusters.
  * @param numIter  Number of iterations.
+ * @param trainingCost Sum of squared distances to the nearest centroid for all points in the
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: Would the cost ever be something besides sum of squares? maybe not, just wondering if we should say here what the cost function is

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala [hunk 4]
```diff
@@ -310,4 +317,6 @@ class BisectingKMeansSummary private[clustering] (
     predictionCol: String,
     featuresCol: String,
     k: Int,
-    numIter: Int) extends ClusteringSummary(predictions, predictionCol, featuresCol, k, numIter)
+    numIter: Int,
+    @Since("3.0.0") val trainingCost: Double)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: I don't think it's a big deal for 3.0, but we lose the constructor without the new param. That's probably OK as the summary kind of needs this value.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala [hunk 5]
```diff
@@ -41,11 +41,12 @@ import org.apache.spark.sql.{Row, SparkSession}
...
 class BisectingKMeansModel private[clustering] (
     private[clustering] val root: ClusteringTreeNode,
-    @Since("2.4.0") val distanceMeasure: String
+    @Since("2.4.0") val distanceMeasure: String,
+    @Since("3.0.0") val trainingCost: Double
   ) extends Serializable with Saveable with Logging {
 
   @Since("1.6.0")
-  def this(root: ClusteringTreeNode) = this(root, DistanceMeasure.EUCLIDEAN)
+  def this(root: ClusteringTreeNode) = this(root, DistanceMeasure.EUCLIDEAN, 0.0)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: On the other hand, we did preserve this old constructor, and that's fine to keep. The other issue I see here is that the cost is 0, when the cost is really unknown.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala [hunk 6]
```diff
@@ -195,7 +200,7 @@ object BisectingKMeansModel extends Loader[BisectingKMeansModel] {
       val data = rows.select("index", "size", "center", "norm", "cost", "height", "children")
       val nodes = data.rdd.map(Data.apply).collect().map(d => (d.index, d)).toMap
       val rootNode = buildTree(rootId, nodes)
-      new BisectingKMeansModel(rootNode, DistanceMeasure.EUCLIDEAN)
+      new BisectingKMeansModel(rootNode, DistanceMeasure.EUCLIDEAN, 0.0)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Is it possible to compute the cost here after load rather than setting it to 0?

**Review by srowen — COMMENTED** [review]
My only real concern is that old models have a training cost of 0, when it's unknown really. I don't think it's worth making the new value an `Option[Double]` because it's not really that optional. If we can compute it in more cases, that's great, would be fine. If not, probably still OK, just less ideal.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 5 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: yes, because this is public, so users may rely on it. The idea is that this is indeed a "new feature" (previously is was not accessible) and we are not guaranteeing new features in the MLLib API. I just followed the same approach which was used for KMeans.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 6 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Despite it would be possible, I don't think it is a good idea, as this is only for the old MLLib API and it would introduce a significant performance overhead (a pass over all the dataset) for an information which may not be useful at all...

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: Would it not just be the same? `rootNode.leafNodes.map(_.cost).sum`? If that cost info is present in the nodes (?) it doesn't need a pass over data (which indeed doesn't exist at this point). If it's valuable enough to include at all, should this info not be correct where it is in fact available?

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: yes, you're right, thanks. It is indeed available. I am doing it, thanks.

**Comment by mgaido91** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala (hunk 6 above)
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: I think it looks OK except for this comment?

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala]: yes, right, sorry, I missed it, I did it only for the version 2.0 and missed this one. I am updating it, thanks.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
